### DISCUSSION TITLE
How to write a DSL including python expressions

### DISCUSSION THREAD

**Discussion by lesire** [textX/textX#338] [Q&A]
Hi,
I am investigating the use of TextX to create a parser for a small DSL whose aim is to describe services provided by autonomous robot systems.
In this DSL, I would like to be able to describe some attributes of an element as a python expression, something like in this simple model:

```
my_element my_element_name {
  parameter p type Position
  guard p.x > 0
}
```

or 

`
guard p.isValid()
`

Is there is simple way to integrate the parsing of these 'guard' attributes? The objective would be to evaluate these guards in the interpretation of my model.
Thanks for your help.

**Comment by goto40** [comment]
Hi.

You will probably need to specify the format of your formulas. Alternatively, you can read them as string, but this is maybe not what you want. 

The examples contain a sample for expressions:

https://github.com/textX/textX/tree/master/examples/expression

You could probably adapt it for your guard expressions. 

Maybe the following unittest may help, if your data structures you refer to (e. g., 'p.x' in your example) can contain each other:

https://github.com/textX/textX/blob/master/tests/functional/examples/test_hierarchical_data_structures_referencing_attributes.py

We would be happy to see your progress...

**Comment by lesire** [comment]
Thanks for your quick response.
I'll look at the expression parser link you mentioned.
It would solve part of the kind of expression I need to parse, but I still need some way to use Python methods or functions... like 
- "p.isValid()", where p would be an object of class Point, whose attributes may be defined from my language, but the method a built-in method from the Point class
- "is_valid(p)", where is_valid would be a function defined in a python module; I just wonder how to link with this module.function when parsing my expression

Thanks again

**Comment by goto40** [comment]
You could create a custom scope provider to include functions and attributes from outside textX (=from the python runtime).

In the following unittest we create a custom scope provider to lookup data from outside textX, from a json file:
https://github.com/textX/textX/blob/master/tests/functional/test_scoping/test_reference_to_nontextx_attribute.py

**Comment by goto40** [comment]
The difficulty (or challenge) is to differentiate between atributes/variables and method/functions. Once we are there, we have to fine tune the grammar to introduce the correct base classes and adjust the order of referenced parser rules.

**Comment by goto40** [comment]
Please have a look at #230 to see how you could reference a python attribute.

Here we add python objects via a custom scope provider.

**Comment by igordejanovic** [comment]
@lesire If you need a full Python expression support than maybe, as @goto40 initially suggested, is better to collect the whole expression as a string and use eval where you would set globals/locals based on the rest of the model.

For example:

```python
In [1]: from textx import metamodel_from_str                                          

In [2]: mm = metamodel_from_str(''' 
   ...: Element: 'my_element' name=ID '{' 
   ...:   parameters+=Parameter 
   ...:   ('guard' guard_exp=/[^\n]+/)? 
   ...: '}'; 
   ...: Parameter: 'parameter' name=ID 'type' type=ID; 
   ...: ''')                                                                          

In [3]: model = mm.model_from_str(''' 
   ...: my_element my_element_name { 
   ...:   parameter p type Position 
   ...:   guard p.x > 0 
   ...: } 
   ...: ''')                                                                          

In [4]: model                                                                         
Out[4]: <Element:my_element_name>

In [5]: model.name                                                                    
Out[5]: 'my_element_name'

In [6]: model.parameters                                                               
Out[6]: [<Parameter:p>]

In [7]: model.guard_exp                                                               
Out[7]: 'p.x > 0'

```

Now, you would need to replace `Parameter` dynamically generated class with your. See [custom classes](http://textx.github.io/textX/stable/metamodel/#custom-classes) on how do do that. Put `model.parameters` elements (which will be instances of your `Parameter` class) into `globals/locals` of `eval` call and evaluate `model.guard_exp`. This is just a small example but should get you started.
